### PR TITLE
L27: C++ API review

### PR DESCRIPTION
**Pull request by vjpai**
Continues the internalization effort started in #28 . Implemented in grpc/grpc#14648 .

Discussion thread at https://groups.google.com/forum/#!topic/grpc-io/LTFUwLhqEHE

Seeking feedback from @makdharma , @chwarr , and any others that care to discuss.

Earliest possible merger is March 30, 2018.

### CODE REVIEW

File: L27-c++-api-review.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+Maintain separation of C++ classes and functions that are intended
+to be directly used by application developers from those that are
+supposed to be used only through codegen or for that exist for gRPC's
+internal implementation (e.g., interactions with the core API).
+
+## Background
+
+[Previous work](https://github.com/grpc/proposal/pull/28) started to
+separate internal-only from public sections of the gRPC C++ API by
+moving selected compoennts to namespace `grpc::internal`. That effort
```
Comment (ericgribkoff) [inline L27-c++-api-review.md]: s/compoennts/components/

File: L27-c++-api-review.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+1. Co-opt items from the core surface API for which there is no other 
+   option, but try to give an alternative and more idiomatically C++
+   version for those when possible
+1. Move classes or functions to `namespace grpc::internal` in some
+   cases
+1. Mark pieces of the API as being meant only for
+   internal/advanced/specialized use as appropriate if typical
+   applications are not expected to use them. These are particularly
+   relevant for convenience libraries that wrap gRPC features but
+   provide their own tuning parameters.
```
[the remark this answers is not in the document]
Comment (chwarr) [reply L27-c++-api-review.md]: @yang-g, if the last two categories go away, how does someone implementing code generation know what APIs are available at that layer? gRFC https://github.com/grpc/proposal/pull/28 and [its implementation](https://github.com/grpc/grpc/pull/11572) moved some things into an "internal" namespace, e.g., `::grpc::ClientReader<R>::internal::Create`. This implies that some parts of `grpc::internal` *are not* implementation details, but part of a lower-level/more advanced API.

**Comment by dfawley** [comment]
For consistency, please rename this file to `L27-cpp-...` before merging.  Thanks!

**Comment by vjpai** [comment]
Abandoned.

**Pull request closed without merging**